src/day8/re_frame/async_flow_fx.cljs

```diff
@@ -19,14 +19,37 @@
       m)
     (dissoc m k)))
 
+
+(defn as-callback-pred
+  "Looks at the required-events items and returns a predicate which
+  will either
+  - match only the event-keyword if a keyword is supplied
+  - match the entire event vector if a collection is supplied
+  - returns a the item itself if a fn is supplied"
+  [callback-pred]
+  (when callback-pred
+    (cond (fn? callback-pred) callback-pred
+          (keyword? callback-pred) (fn [[event-id _]]
+                                     (= callback-pred event-id))
+          (coll? callback-pred) (fn [event-v]
+                                  (= callback-pred event-v))
+          :else (throw
+                  (ex-info (str (pr-str callback-pred)
+                             " isn't an event predicate")
+                    {:callback-pred callback-pred})))))
+
 (defn seen-all-of?
   [required-events seen-events]
-  (empty? (set/difference required-events seen-events)))
+  (every?
+    (fn [pred] (some (fn [e] (pred e)) seen-events))
+    (map as-callback-pred required-events)))
 
 
 (defn seen-any-of?
   [required-events seen-events]
-  (some? (seq (set/intersection seen-events required-events))))
+  (some? (some
+           (fn [pred] (some (fn [e] (pred e)) seen-events))
```

Comment:
Can `(some (fn [e] (pred e)) seen-events)` be written as `(some pred seen-events)`? Perhaps it would also be easier to read if it was extracted into a let binding as `any-pred?`